File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/java/exercises/practice/state-of-tic-tac-toe/src/main/java/StateOfTicTacToe.java
enum GameState {
    WIN, DRAW, ONGOING
}

class StateOfTicTacToe {
    public GameState determineState(String[] board) {
        int xCount = 0;
        int oCount = 0;
        
        // Count X and O marks
        for (String row : board) {
            for (char c : row.toCharArray()) {
                if (c == 'X') xCount++;
                else if (c == 'O') oCount++;
            }
        }
        
        // Validate turn order (X starts and alternates)
        if (xCount != oCount && xCount != oCount + 1) {
            throw new IllegalArgumentException("Invalid turn order");
        }
        
        boolean xWins = hasWin(board, 'X');
        boolean oWins = hasWin(board, 'O');
        
        // Validate only one possible winner
        if (xWins && oWins) {
            throw new IllegalArgumentException("Multiple winners");
        }
        
        // Check for continued play after win
        if (xWins) {
            if (xCount != oCount + 1) {
                throw new IllegalArgumentException("Game continued after X win");
            }
            return GameState.WIN;
        }
        
        if (oWins) {
            if (xCount != oCount) {
                throw new IllegalArgumentException("Game continued after O win");
            }
            return GameState.WIN;
        }
        
        // Determine final state
        return (xCount + oCount) == 9 ? GameState.DRAW : GameState.ONGOING;
    }
    
    private boolean hasWin(String[] board, char player) {
        // Check rows and columns
        for (int i = 0; i < 3; i++) {
            if (board[i].charAt(0) == player && board[i].charAt(1) == player && board[i].charAt(2) == player) {
                return true;
            }
            if (board[0].charAt(i) == player && board[1].charAt(i) == player && board[2].charAt(i) == player) {
                return true;
            }
        }
        
        // Check diagonals
        return (board[0].charAt(0) == player && board[1].charAt(1) == player && board[2].charAt(2) == player)
            || (board[0].charAt(2) == player && board[1].charAt(1) == player && board[2].charAt(0) == player);
    }
}

File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/javascript/exercises/practice/state-of-tic-tac-toe/state-of-tic-tac-toe.js
export const gamestate = (board) => {
  let xCount = 0, oCount = 0;
  for (const row of board) {
    for (const cell of row) {
      if (cell === 'X') xCount++;
      else if (cell === 'O') oCount++;
    }
  }

  if (xCount !== oCount && xCount !== oCount + 1) {
    if (oCount > xCount) {
      throw new Error('Wrong turn order: O started');
    } else {
      throw new Error('Wrong turn order: X went twice');
    }
  }

  const hasWon = (player) => {
    // Check rows
    for (const row of board) {
      if (row[0] === player && row[1] === player && row[2] === player) {
        return true;
      }
    }

    // Check columns
    for (let col = 0; col < 3; col++) {
      if (board[0][col] === player && board[1][col] === player && board[2][col] === player) {
        return true;
      }
    }

    // Check diagonals
    if (board[0][0] === player && board[1][1] === player && board[2][2] === player) {
      return true;
    }
    if (board[0][2] === player && board[1][1] === player && board[2][0] === player) {
      return true;
    }

    return false;
  };

  const xWon = hasWon('X');
  const oWon = hasWon('O');

  if (xWon && oWon) {
    throw new Error('Impossible board: game should have ended after the game was won');
  }

  if (xWon && xCount !== oCount + 1) {
    throw new Error('Impossible board: game should have ended after the game was won');
  }

  if (oWon && xCount !== oCount) {
    throw new Error('Impossible board: game should have ended after the game was won');
  }

  const isBoardFull = () => {
    for (const row of board) {
      for (const cell of row) {
        if (cell === ' ') {
          return false;
        }
      }
    }
    return true;
  };

  if (xWon || oWon) {
    return 'win';
  } else if (isBoardFull()) {
    return 'draw';
  } else {
    return 'ongoing';
  }
};
